kernel/src/net/socket/netlink/table.rs

```diff
@@ -0,0 +1,218 @@
+// SPDX-License-Identifier: MPL-2.0
+
+use alloc::collections::BTreeMap;
+
+use super::{
+    addr::{NetlinkProtocolId, NetlinkSocketAddr, PortNum},
+    multicast_group::{GroupIdSet, MulticastGroup, MAX_GROUPS},
+};
+use crate::{net::socket::netlink::addr::UNSPECIFIED_PORT, prelude::*, util::random::getrandom};
+
+pub static NETLINK_SOCKET_TABLE: NetlinkSocketTable = NetlinkSocketTable::new();
+
+/// All bound netlink sockets.
+pub struct NetlinkSocketTable {
+    protocols: RwMutex<BTreeMap<NetlinkProtocolId, RwMutex<ProtocolSocketTable>>>,
+}
+
+impl NetlinkSocketTable {
+    pub const fn new() -> Self {
+        Self {
+            protocols: RwMutex::new(BTreeMap::new()),
+        }
+    }
+
+    /// Adds a new netlink protocol.
+    fn add_new_protocol(&self, protocol: NetlinkProtocolId) {
+        let mut protocols = self.protocols.write();
+        if protocols.contains_key(&protocol) {
+            return;
+        }
+        let new_protocol = RwMutex::new(ProtocolSocketTable::new(protocol));
+        protocols.insert(protocol, new_protocol);
+    }
+
+    pub fn bind(
+        &self,
+        protocol: NetlinkProtocolId,
+        addr: &NetlinkSocketAddr,
+    ) -> Result<BoundHandle> {
+        let protocols = self.protocols.read();
+
+        let Some(protocol_sockets) = protocols.get(&protocol) else {
+            return_errno_with_message!(Errno::EINVAL, "the netlink protocol does not exist")
+        };
+
+        let mut protocol_sockets = protocol_sockets.write();
+        protocol_sockets.bind(addr)
+    }
+}
+
+/// Bound socket table of a single netlink protocol.
+///
+/// Each table can have bound sockets for unicast
+/// and at most 32 groups for multicast.
+struct ProtocolSocketTable {
+    id: NetlinkProtocolId,
+    unicast_sockets: BTreeSet<PortNum>,
+    multicast_groups: Box<[MulticastGroup]>,
```

Comment:
`unicast_sockets` and `multicast_groups` should be a `port number -> socket` map so that the kernel can deliver the message to it. The current form is wired. We should at least have a TODO?